### PR TITLE
GH-100762: Don't call `gen.throw()` in `gen.close()`, unless necessary.

### PR DESCRIPTION
**Pull request by markshannon**
Change `oparg` of `YIELD_VALUE` to hold the exception handler depth.
We can then test that in `gen.close()` to avoid calling `gen.throw()` unless necessary.

* Issue: gh-100762

### CODE REVIEW

**Comment by markshannon** [comment]
Using the benchmarks in https://github.com/python/cpython/pull/101011#issue-1532104206, but using `timeit` instead of `pyperf` and a non-pgo and no-lto build because I'm impatient.
I get no significant change on "comprehension", "generator exhaust", and "all", but runtime is roughly halved for "generator" and "all with comparison"

File: Lib/importlib/_bootstrap_external.py [hunk 1]
```diff
@@ -430,6 +430,8 @@ def _write_atomic(path, data, mode=0o666):
 #     Python 3.12a1 3514 (Remove ASYNC_GEN_WRAP, LIST_TO_TUPLE, and UNARY_POSITIVE)
 #     Python 3.12a1 3515 (Embed jump mask in COMPARE_OP oparg)
 
+#     Python 3.12a1 3517 (Change YIELD_VALUE oparg to exception block depth)
```
Comment (brandtbucher) [inline Lib/importlib/_bootstrap_external.py]: I'm guessing you have a 3516 PR open somewhere?

**Review by brandtbucher — APPROVED** [review]

File: Doc/library/dis.rst [hunk 2]
```diff
@@ -699,8 +699,8 @@ iterations of the loop.
 
    Yields ``STACK.pop()`` from a :term:`generator`.
 
-    .. versionchanged:: 3.11
-       oparg set to be the stack depth, for efficient handling on frames.
+    .. versionchanged:: 3.12
+       oparg set to be the exception block depth, for efficient closing of generators.
```
Comment (gvanrossum) [inline Doc/library/dis.rst]: Shouldn't the 3.11 note be retained (assuming it's truthful)?

File: Objects/genobject.c [hunk 3]
```diff
@@ -354,15 +354,37 @@ gen_close(PyGenObject *gen, PyObject *args)
     PyObject *yf = _PyGen_yf(gen);
     int err = 0;
 
+    if (gen->gi_frame_state == FRAME_CREATED) {
+        gen->gi_frame_state = FRAME_COMPLETED;
+        Py_RETURN_NONE;
```
Comment (gvanrossum) [inline Objects/genobject.c:359]: You _could_ drop this line and everything should work the same. Probably about the same speed too (the memory locations read on the next line will be in the L1 cache).

File: Python/compile.c [hunk 4]
```diff
@@ -7332,6 +7329,9 @@ label_exception_targets(basicblock *entryblock) {
                 }
             }
             else {
+                if (instr->i_opcode == YIELD_VALUE) {
+                    instr->i_oparg = except_stack->depth;
+                }
```
Comment (gvanrossum) [inline Python/compile.c:7334]: I'm just trusting that this is the right place to insert this. @iritkatriel ?

File: Objects/genobject.c [hunk 5]
```diff
@@ -354,15 +354,37 @@ gen_close(PyGenObject *gen, PyObject *args)
...
         err = gen_close_iter(yf);
         gen->gi_frame_state = state;
         Py_DECREF(yf);
     }
-    if (err == 0)
+    _PyInterpreterFrame *frame = (_PyInterpreterFrame *)gen->gi_iframe;
+    /* It is possible for the previous instruction to not be a
+     * YIELD_VALUE if the debugger has changed the lineno. */
+    if (err == 0 && frame->prev_instr->opcode == YIELD_VALUE) {
+        assert(frame->prev_instr[1].opcode == RESUME);
```
Comment (gvanrossum) [inline Objects/genobject.c:375]: Is this assert just here so we'll trip it if the compiler ever changes?

File: Lib/importlib/_bootstrap_external.py [hunk 6]
```diff
@@ -430,6 +430,7 @@ def _write_atomic(path, data, mode=0o666):
 #     Python 3.12a1 3514 (Remove ASYNC_GEN_WRAP, LIST_TO_TUPLE, and UNARY_POSITIVE)
 #     Python 3.12a1 3515 (Embed jump mask in COMPARE_OP oparg)
 #     Python 3.12a1 3516 (Add COMAPRE_AND_BRANCH instruction)
```
Comment (gvanrossum) [inline Lib/importlib/_bootstrap_external.py]: Maybe fix the typo here (COMAPRE -> COMPARE)?

**Review by gvanrossum — COMMENTED** [review]
I guess this will not be easy to backport, as we won't want to change the meaning of oparg in 3.11? Then again it appears to be unused except for the assert. Maybe it should just be a bit indicating whether it's safe to optimize out the throw?

**Comment by brandtbucher** [comment]
> I guess this will not be easy to backport, as we won't want to change the meaning of oparg in 3.11?

I think a similar change in 3.11 would actually be simpler: just check if an exception table exists. If not, don't throw.

The only reason we need the `YIELD_VALUE` oparg in this 3.12 patch is because every generator now has an implicit `try`/`except` around the entire body, so it's a bit trickier to figure out if we're in that handler (which we can safely ignore) or something else.

**Comment by gvanrossum** [comment]
Great, it'll just be *different*. :-)

File: Objects/genobject.c (hunk 5 above)
Comment (markshannon) [reply Objects/genobject.c:375]: Yes. Plus an extra assert never does any harm.

File: Objects/genobject.c (hunk 3 above)
Comment (markshannon) [reply Objects/genobject.c:359]: I think it is more readable this way, so I'll leave it.

File: Python/compile.c (hunk 4 above)
Comment (iritkatriel) [reply Python/compile.c:7334]: I can't think of a better place. We don't have the except stack depth anywhere else.

File: Lib/importlib/_bootstrap_external.py (hunk 1 above)
Comment (markshannon) [reply Lib/importlib/_bootstrap_external.py]: Already merged

File: Doc/library/dis.rst (hunk 2 above)
Comment (markshannon) [reply Doc/library/dis.rst]: done

**Comment by brandtbucher** [comment]
I don’t think this should be backported as-is, right? (See my comment above.)

**Comment by markshannon** [comment]
@brandtbucher agreed.
The change to genobject.c need backporting (with the change you suggest for determining when to skip the call to `gen.throw()`, but the other changes are 3.12 specific.

**Pull request merged**